### ISSUE TITLE
the problem for s3read can't  be fixed 

### ISSUE THREAD

**Issue by Bee-zest** [robertzk/s3mpi#72]
> library(s3mpi)
 > data_train <- s3read(name = "data_train_extra_features.RDS",path =  "s3://[redacted]/Blackwell/shared/Train/")

 > sh: 1: : Permission denied Error in gzfile(file, "rb") : cannot open the connection In addition: Warning messages: 1: In fetch(path, storage_format, bucket_location, ...) :   Nothing exists for key s3://<redacted>/data_train_extra_features.RDS 2: In gzfile(file, "rb") :   cannot open compressed file '/tmp/RtmploRNgS/file194aa63c626be', probable reason 'No such file or directory'

**Comment by peterhurford** [comment]
This is the error that you get when you read a key that does not exist. Are you reading the correct key? How did you `s3store` the data?

@robertzk @kirillseva We may want to make this kind of "key not found" error more explicit.

**Comment by Bee-zest** [comment]
My colleague may install the old version. For me, it always get this error.
When I use s3.get to read data from s3, it works. But for s3read, it aways
get this error.

**Comment by Bee-zest** [comment]
For the simplest exmaple, it still does not work

**Comment by peterhurford** [comment]
@Bee-zest So `s3mpi:::s3.get("s3://spotloan-marketing/Blackwell/shared/Train/data_train_extra_features.RDS", cache = FALSE)` works for you? How do you use `s3.get` to get your data?

**Comment by robertzk** [comment]
Thanks @Bee-zest -- good point, I get a similar error.

```
R-ENGINES> s3read('idontexist')
ERROR: S3 error: 404 (Not Found):
```

Historically we've never minded this, but given that we're trying to improve the public-facing aspects of these packages it makes sense to wrap this with a nicer error.

**Comment by peterhurford** [comment]
@robertzk I expanded on that and filed https://github.com/robertzk/s3mpi/issues/73

**Comment by Bee-zest** [comment]
No, the same error with s3read. The

data_train <-
s3.get("s3://spotloan-marketing/Blackwell/production/Train/2017-06-30_Blackwell.RDS")
works for me.

**Comment by robertzk** [comment]
@Bee-zest What if you set your s3 buckets:

```r
options(s3mpi.path = "s3://spotloan-marketing/")
data_train <- s3read("Blackwell/production/Train/2017-06-30_Blackwell.RDS")
```

**Comment by Bee-zest** [comment]
Got this error

[image: Inline image 2]

**Comment by robertzk** [comment]
Thanks @Bee-zest -- can you post in the GitHub URL rather than email? The image doesn't seem to have been attached.

https://github.com/robertzk/s3mpi/issues/72

**Comment by Bee-zest** [comment]
> options(s3mpi.path = "s3://spotloan-marketing/")
> data_train <- s3read("Blackwell/production/Train/2017-06-30_Blackwell.RDS")


ERROR: Parameter problem: Expecting S3 URI instead of 'Blackwell/production/Train/2017-06-30_Blackwell.RDS'
 Show Traceback
 
 Rerun with Debug
 Error in gzfile(file, "rb") : cannot open the connection In addition: Warning messages:
1: running command 's3cmd get Blackwell/production/Train/2017-06-30_Blackwell.RDS /tmp/RtmploRNgS/file194aa7d15953 --bucket-location US --no-progress  ' had status 1 
2: In gzfile(file, "rb") :
  cannot open compressed file '/tmp/RtmploRNgS/file194aa7d15953', probable reason 'No such file or directory'

**Comment by robertzk** [comment]
@Bee-zest Strange, it doesn't seem to be using the path. Try:

```r
data_train <- s3read("Blackwell/production/Train/2017-06-30_Blackwell.RDS", "s3://spotloan-marketing/")

**Comment by Bee-zest** [comment]
The same error.. Hi robertzk. Do you have old version?

**Comment by peterhurford** [comment]
@Bee-zest What version are you on? We're the developers, so naturally we have the latest version. ;)

**Comment by Bee-zest** [comment]
yes. I think I'm using the latest version. I installed the packages from your github homepage yesterday.

**Comment by Bee-zest** [comment]
Hi, could you please send me the old version of s3mpi? I can't fix the
problem of the latest version

**Comment by Bee-zest** [comment]
s3read <- function(name = NULL, .path = s3path(), cache = TRUE, ...) {
  if (is.null(name)) name <- grab_latest_file_in_s3_dir(.path)

  if (substr(.path, nchar(.path), nchar(.path)) != "/") { .path <-
paste0(.path, "/") }
  s3key <- paste(.path, name, sep = '')

  if (!isTRUE(cache) || is.null(getOption('s3mpi.cache'))) {
    value <- s3.get(s3key, ...)
  } else if (is.not_cached(value <- s3cache(s3key))) {
    value <- s3.get(s3key, ...)
    s3cache(s3key, value)
  }
  s3normalize(value, TRUE)
}



options(s3mpi.path = 's3://spotloan-marketing/Blackwell/production/Train/')

data_train1 <- s3read(name = "2017-06-30_Blackwell.RDS")

nrow(data_train1)



I check your previous function for s3read. It now works. But why the latest
one doesn't work

**Comment by robertzk** [comment]
Hey @Bee-zest - which version are you looking for? You can pull old versions from github by grabbing the associated tag.

```r 
git clone git@github.com:robertzk/s3mpi.git
cd s3mpi
git checkout <version>
Rscript --vanilla -e 'devtools::install(".")'
```

**Comment by robertzk** [comment]
Thanks @Bee-zest -- it appears the rocco documentation is very out of date. The below call should install the version as of that rocco doc.

```r
devtools::install_github("robertzk/s3mpi@0.2.12")
```

**Comment by Bee-zest** [comment]
Thanks Robert. I install the version 0.2.12. The same error.

**Comment by Bee-zest** [comment]
The 0.2.12 doesn't work
